### PR TITLE
Overflow underflow detection

### PR DESCRIPTION
**Pull request by qingxinhome**
## What type of PR is this?

- [ ] API-change
- [ ] BUG
- [x] Improvement
- [ ] Documentation
- [ ] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue #3440

## What this PR does / why we need it:

Rework arithmetic operations and add overflow detection

### CODE REVIEW

**Review by fengttt — APPROVED** [review]

File: cgo/decimal.c [hunk 1]
```diff
@@ -585,27 +585,85 @@ int32_t Decimal128_DivInt64(int64_t *r, int64_t *a, int64_t b)
...
+{
+    decQuad tmp;
+    DECLARE_DEC128_CTXT;
+    decQuadSubtract(&tmp, DecQuadPtr(a), DecQuadPtr(b), &_fn_dc);
+    decQuadReduce(DecQuadPtr(r), &tmp, &_fn_dc);
+//    CHECK_OFUF;
+    return RC_SUCCESS;
+}
+
+int32_t Decimal128_MulNoCheck(int64_t *r, int64_t *a, int64_t *b)
```
Comment (fengttt) [inline cgo/decimal.c:625]: Let's always supply a context.

File: pkg/vectorize/div/dec.go [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+		flag |= LEFT_IS_SCALAR
+	}
+	if ys.IsScalar() {
+		flag |= RIGHT_IS_SCALAR
+	}
+
+	rc := C.Decimal64_VecDiv(dec128PtrToC(&rt[0]), dec64PtrToC(&xt[0]), dec64PtrToC(&yt[0]),
+		C.uint64_t(len(rt)), (*C.uint64_t)(nulls.Ptr(rs.Nsp)), C.int32_t(flag))
+	if rc != 0 {
+		return moerr.NewError(moerr.OUT_OF_RANGE, "Decimal64 div overflow")
```
[the remark this answers is not in the document]
Comment (qingxinhome) [reply pkg/vectorize/div/dec.go:55]: OK, I will sort out and refine the error information

File: cgo/decimal.c (hunk 1 above)
Comment (qingxinhome) [reply cgo/decimal.c:625]: OK, no problem

File: Makefile [hunk 3]
```diff
@@ -115,7 +115,7 @@ pb: vendor-build generate-pb fmt
 RACE_OPT := 
 DEBUG_OPT := 
 CGO_DEBUG_OPT :=
-CGO_OPTS=CGO_CFLAGS="-I$(ROOT_DIR)/cgo" CGO_LDFLAGS="-L$(ROOT_DIR)/cgo -lmo"
+CGO_OPTS=CGO_CFLAGS="-I$(ROOT_DIR)/cgo" CGO_LDFLAGS="-L$(ROOT_DIR)/cgo -lmo -lm"
```
Comment (fengttt) [inline Makefile:118]: You need to find out all the -lmo and update it.   I think somewhere in our code, we still have cgo instruction in go file.

Need to update everything

File: cgo/arith.c [hunk 4]
```diff
@@ -110,26 +110,28 @@
...
-    TGT = (A) * (B);                                 \
-    opflag = ((A) != 0) && (((TGT) / (A)) != B)
-
-#define MUL_SIGNED_OVFLAG_CHECK                      \
-    if (opflag != 0)    {                            \
-        return RC_OUT_OF_RANGE;                      \
+#define MUL_SIGNED_OVFLAG(TGT, A, B, MAXVAL, MINVAL, ZT ,UPTYPE)               \
+    temp = (UPTYPE)(A) * (UPTYPE)(B);                                          \
+    TGT = (ZT)temp;                                                            \
+    opflag = ((A ^ B) > 0 && temp > MAXVAL) || ((A ^ B) < 0 && temp < MINVAL)
```
Comment (fengttt) [inline cgo/arith.c:116]: You don't need to test A^B.    Ust test temp < MIN || temp > MAX.

**Review by daviszhen — APPROVED** [review]

**Pull request merged**